FirebaseAuth to Swift

paulb777 opened the pull request:
Experimental PR to convert FirebaseAuth to Swift starting with the AuthProviders.

The library builds with `pod gen` and the unit tests run with @ncooke3's mixed language Swift Package Manager PR.

This is very much a work in progress, but feel free to check it out and send PR's to the pb-auth-swift branch

paulb777 commented on FirebaseAuth.podspec:
```diff
@@ -47,7 +48,7 @@ supports email and password accounts, as well as several 3rd party authenticatio
   s.pod_target_xcconfig = {
     'GCC_C_LANGUAGE_STANDARD' => 'c99',
-    'HEADER_SEARCH_PATHS' => '"${PODS_TARGET_SRCROOT}"'
+    'HEADER_SEARCH_PATHS' => '"${PODS_TARGET_SRCROOT}" "${OBJECT_FILE_DIR_normal}/${NATIVE_ARCH_ACTUAL}"'
```
find FirebaseAuth-Swift.h from a pod gen project

paulb777 commented on FirebaseAuth/Sources/Auth/FIRAuth.m:
```diff
@@ -28,16 +28,12 @@
 #import "FirebaseAuth/Sources/Auth/FIRAuthDispatcher.h"
 #import "FirebaseAuth/Sources/Auth/FIRAuthGlobalWorkQueue.h"
 #import "FirebaseAuth/Sources/Auth/FIRAuthOperationType.h"
-#import "FirebaseAuth/Sources/AuthProvider/Email/FIREmailPasswordAuthCredential.h"
```
Internal Objective C headers need be public to be seen by Swift via `pod gen`

paulb777 commented on FirebaseAuth/Sources/Auth/FIRAuth.m:
```diff
@@ -820,8 +821,8 @@ - (void)signInAndRetrieveDataWithGameCenterCredential:(FIRGameCenterAuthCredenti
                                                }
                                                FIRAdditionalUserInfo *additionalUserInfo =
                                                    [[FIRAdditionalUserInfo alloc]
-                                                       initWithProviderID:
```
A breaking change to consider when we move to Swift implementation. Global strings are problematic, especially across languages.

paulb777 commented on FirebaseAuth/Sources/Auth/FIRAuth_Internal.h:
```diff
@@ -32,12 +32,6 @@ NS_ASSUME_NONNULL_BEGIN
-    @brief The configuration object comprising of paramters needed to make a request to Firebase
-        Auth's backend.
- */
-@property(nonatomic, copy, readonly) FIRAuthRequestConfiguration *requestConfiguration;
```
Moved to FIRAuth.h to be visible from Swift

paulb777 commented on FirebaseAuth/Tests/Sample/Sample/FacebookAuthProvider.m:
```diff
@@ -48,7 +48,7 @@ - (void)getAuthCredentialWithPresentingViewController:(UIViewController *)viewCo
 
   [ApplicationDelegate setOpenURLDelegate:self];
-  [FBSDKSettings setAppID:kFacebookAppID];
+  //[FBSDKSettings setAppID:kFacebookAppID];
```
fix me

mortenbekditlevsen commented:
Hi Paul,
I assume that this only builds using CocoaPods since SwiftPM doesn't handle mixed language targets, right?
In my own similar experiments I've created a Swift target (something like FirebaseAuthSwift) which the Objective-C FirebaseAuth target then depends on. 
I would love to contribute, if I can. Can you share more about how you would like to the project to progress? Do you want to keep 100% Objective-C compatibility for the sake of testing? Or would you like tests to be converted too, so that in time, many of the internals can be hidden away from Objective-C consumers too?

Really exciting stuff! :-)

paulb777 commented:
Hi Morten,

Thanks for reaching out. Yes, we would like to convert the Firebase implementation and tests to Swift. At least for now, we'd like to continue to support the Objective-C API as well, although at some point, we expect that ObjC support will get too costly and that asking ObjC users to use old versions will be a mitigation. One exception area might be to immediately drop support for globally scoped strings from the API, which comes up in this PR for the AuthProvider ids.

So far this is an exploratory project to test the new SPM mixed language support and to experiment with Swift migration strategies for Firebase. Until Nick's PR releases with Xcode, development is a little tricky since I need to use CocoaPods to develop in Xcode, but I don't get the advantages of Nick's PR for making internal headers visible to Swift, so I need to move a bunch of headers to Public.  Nick's PR is very useful in enabling me to run the Objective-C unit tests to test the Swift implementations.

Because Auth's use of properties across classes is spaghetti-ish, I see the mixed language approach as currently only a useful porting tool and not releasable. Alternatively, we could also do more of a cleaner componentized redesign in which a mixed language approach could work. However, while Auth is substantially bigger than Storage and Functions, it's still probably small enough, that we might as well convert the whole thing and then maybe restructure.

mortenbekditlevsen commented on FirebaseAuth/Sources/Swift/AuthProvider/EmailAuthProvider.swift:
```diff
@@ -0,0 +1,92 @@
+  @objc public let password: String?
+  @objc public let link: String?
+
+  @objc public init(withEmail email: String, password: String) {
```
With regards to obj-c naming, `init(email: String, password: String)` will translate to `initWithEmail:password:` too. My feeling is that the `with` prefix is an objc-ism, so it can be skipped in the Swift version of the api.

mortenbekditlevsen commented:
Hi @ncooke3 and @paulb777 ,
That's a really nice proposal - and one that would help many projects transitioning to Swift.

I tried building your nc/support-mixed-targets branch.
After building I wanted to try running the unit tests in pb-auth-swift.

I'm running:
`../swift-package-manager/.build/arm64-apple-macosx/debug/swift-test`
(where your proposal pr branch is built)
from the root of a checkout of this pr branch. But I get the following output:
```
Fetching https://github.com/erikdoe/ocmock.git from cache
Fetched https://github.com/erikdoe/ocmock.git (0.74s)
Fetching https://github.com/google/gtm-session-fetcher.git from cache
Fetching https://github.com/google/GoogleUtilities.git from cache
Fetching https://github.com/google/promises.git from cache
Fetched https://github.com/google/promises.git (1.04s)
Fetched https://github.com/google/GoogleUtilities.git (1.14s)
Fetched https://github.com/google/gtm-session-fetcher.git (1.14s)
Computing version for https://github.com/google/gtm-session-fetcher.git
Computed https://github.com/google/gtm-session-fetcher.git at 3.1.0 (0.04s)
Computing version for https://github.com/google/GoogleUtilities.git
Computed https://github.com/google/GoogleUtilities.git at 7.11.0 (0.03s)
Computing version for https://github.com/google/promises.git
Computed https://github.com/google/promises.git at 2.0.0 (0.03s)
Creating working copy for https://github.com/erikdoe/ocmock.git
Working copy of https://github.com/erikdoe/ocmock.git resolved at c5eeaa6dde7c308a5ce48ae4d4530462dd3a1110
Creating working copy for https://github.com/google/promises.git
Working copy of https://github.com/google/promises.git resolved at 2.0.0
Creating working copy for https://github.com/google/GoogleUtilities.git
Working copy of https://github.com/google/GoogleUtilities.git resolved at 7.11.0
Creating working copy for https://github.com/google/gtm-session-fetcher.git
Working copy of https://github.com/google/gtm-session-fetcher.git resolved at 3.1.0
error: 'firebase-ios-sdk2': target at '/Users/mortenbekditlevsen/PrivateProjects/firebase-ios-sdk2/FirebaseAuth/Sources' contains mixed language source files; feature not supported until tools version XX
```

May I ask how you are running the unit tests with the custom SwiftPM build?

mortenbekditlevsen commented:
I'm sorry about the off-topic question. I was looking at my own FirebaseAuth -> Swift conversion attempt, which I haven't been looking at for a few months - in order to see if there are bits and pieces that I could contribute with.

I have been developing using SwiftPM, but today it appears that some underlying dependencies was pulled, so I can't open my old project. 

No worries, I'll just move to a more recent branch and try my changes there.

But currently I can't build the project using SwiftPM on either `master` or the latest `10.3.0` tag.

I get the following SwiftPM error message on `master`:

```
Failed to resolve dependencies Dependencies could not be resolved because no versions of 'googleappmeasurement' match the requirement 10.4.0 and root depends on 'googleappmeasurement' 10.4.0.
```

I tried hacking around this by using the `main` branch of `GoogleUtilities` like it's done with the `FIREBASECI_USE_LATEST_GOOGLEAPPMEASUREMENT` setting, but then I get errors when importing `GoogleUtilities_NSData` from `FirebaseCoreInternal`.

Please don't spend much time on this, but might you have a quick pointer to building using SwiftPM?

paulb777 commented:
@mortenbekditlevsen The GoogleAppMeasurement is a known master issue while we're staging issues. You found the right workaround. There are several issues with the latest SPM and current Package.swift. We deferred solving them in favor of a Package.swift in this PR that only builds FirebaseAuth.

The version in check in the SPM branch needs to be commented out as well as one entire file:

~/github/swift-package-manager  $ git status
HEAD detached at c3b732ad2
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   Sources/PackageLoading/TargetSourcesBuilder.swift
	modified:   Tests/SourceControlTests/GitRepositoryTests.swift

no changes added to commit (use "git add" and/or "git commit -a")
~/github/swift-package-manager  $ git diff
diff --git a/Sources/PackageLoading/TargetSourcesBuilder.swift b/Sources/PackageLoading/TargetSourcesBuilder.swift
index 5ca27d8e4..cb5fec681 100644
--- a/Sources/PackageLoading/TargetSourcesBuilder.swift
+++ b/Sources/PackageLoading/TargetSourcesBuilder.swift
@@ -182,9 +182,9 @@ public struct TargetSourcesBuilder {

         // It's an error to contain mixed language source files.
         // FIXME(ncooke3): Update with next version of SPM.
-        if sources.containsMixedLanguage, toolsVersion < .vNext  {
-            throw Target.Error.mixedSources(targetPath)
-        }
+//        if sources.containsMixedLanguage, toolsVersion < .vNext  {
+//            throw Target.Error.mixedSources(targetPath)
+//        }

         return (sources, resources, headers, ignored, others)
     }
diff --git a/Tests/SourceControlTests/GitRepositoryTests.swift b/Tests/SourceControlTests/GitRepositoryTests.swift
index 933f94abc..e009f8c85 100644
--- a/Tests/SourceControlTests/GitRepositoryTests.swift
+++ b/Tests/SourceControlTests/GitRepositoryTests.swift
@@ -1,739 +1,739 @@
-//===----------------------------------------------------------------------===//

paulb777 commented:
Set the `package-path` option to the swift-test scheme to run the FirebaseAuth unit tests after building the `SwiftPM-Package` scheme.

<img width="912" alt="Screen Shot 2023-01-15 at 2 28 22 PM" src="https://user-images.githubusercontent.com/73870/212573347-db94149f-438a-4896-ad71-6a0233038d67.png">

paulb777 commented on FirebaseAuth/Sources/Swift/AuthProvider/AuthProviderStrings.swift:
```diff
@@ -0,0 +1,25 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
```
This file is a proposed new API to address 9236. It needs to duplicate the strings in the provider classes, because String enums are not accessible from ObjC

paulb777 commented:
I'm going to merge this into a new `auth-swift` branch and manage subsequent work via separate PRs.
